src/layer/gridsample.cpp

```diff
@@ -0,0 +1,528 @@
+{
+    float A = -0.75;
+
+    float x1 = t;
+    coeffs[0] = cubic_convolution2(x1 + 1.0, A);
+    coeffs[1] = cubic_convolution1(x1, A);
+
+    // opposite coefficients
+    float x2 = 1.0 - t;
+    coeffs[2] = cubic_convolution1(x2, A);
+    coeffs[3] = cubic_convolution2(x2 + 1.0, A);
+}
+
+static float cubic_interp1d(float x0, float x1, float x2, float x3, float t)
+{
+    float coeffs[4];
+    get_cubic_upsample_coefficients(coeffs, t);
+
+    return x0 * coeffs[0] + x1 * coeffs[1] + x2 * coeffs[2] + x3 * coeffs[3];
+}
+
+int GridSample::forward(const std::vector<Mat>& bottom_blobs, std::vector<Mat>& top_blobs, const Option& opt) const
+{
+    const Mat& bottom_blob = bottom_blobs[0];
+    const Mat& grid = bottom_blobs[1];
+    Mat& top_blob = top_blobs[0];
+
+    int w = bottom_blob.w;
+    int h = bottom_blob.h;
+    int d = bottom_blob.d;
+    int channels = bottom_blob.c;
+    int dims = bottom_blob.dims;
+    size_t elemsize = bottom_blob.elemsize;
+
+    if (dims == 3)
+    {
+        int outw = grid.h;
+        int outh = grid.c;
+
+        top_blob.create(outw, outh, channels, elemsize, opt.blob_allocator);
+        if (top_blob.empty())
+            return -100;
+        if (resize_type == 1) // bilinear
+        {
+            #pragma omp parallel for num_threads(opt.num_threads) // collapse(2)
+            for (int row = 0; row < outh; row++)
+            {
+                for (int col = 0; col < outw; col++)
+                {
+                    // const float* gridptr = grid.depth(row).row(col);
+                    const float* gridptr = grid.channel(row).row(col);
+
+                    // get the coordinate of every output point
+                    float ix = gridptr[0];
+                    float iy = gridptr[1];
+
+                    ix = get_coord(ix, w, padding_mode, align_corner);
+                    iy = get_coord(iy, h, padding_mode, align_corner);
+
+                    // for 3d, we used north-east-south-west
+                    int xnw = static_cast<int>(floor(ix));
+                    int xne = xnw + 1;
+                    int xsw = xnw;
+                    int xse = xne;
+
+                    int ynw = static_cast<int>(floor(iy));
+                    int yne = ynw;
+                    int ysw = ynw + 1;
+                    int yse = ysw;
+
+                    // get the coeff of every output point
+                    float fnw = (xse - ix) * (yse - iy);
+                    float fne = (ix - xsw) * (ysw - iy);
+                    float fsw = (xne - ix) * (iy - yne);
+                    float fse = (ix - xnw) * (iy - ynw);
+
+                    for (int q = 0; q < channels; q++)
+                    {
+                        const float* ptr = bottom_blob.channel(q);
+                        float* outptr = top_blob.channel(q);
+
+                        float ans = 0.f;
+
+                        if (in_bounds(xnw, ynw, w, h))
```

Comment:
get_coord() itself is capable for out-of-bound coord resovling.
padding_mode is not regarded for the three neighbor pixels here and is always treated as zero when out-of-bound. it shall not be like this way.